### PR TITLE
Implement PAYG notifications as a reminder for expiration time 

### PR DESCRIPTION
**Pull request by mariospr**
Note: This PR also includes a fixup for a previous commit.

https://phabricator.endlessm.com/T21848

### CODE REVIEW

File: js/misc/paygManager.js [hunk 1]
```diff
@@ -60,6 +63,25 @@ const DBusErrorsMapping = {
...
+
+// This list defines the different instants in time where we would
+// want to show notifications to the user reminding that the payg
+// subscription will be expiring soon.
+//
+// It contains a list of integers representing the number of seconds
+// earlier to the expiration time when we want to show a notification,
+// which needs to be sorted in DESCENDING order.
+const notificationAlertTimesSecs = [
+    60 * 60 * 2, // 2 hours
```
Comment (rshuler) [inline js/misc/paygManager.js]: I think that this first warning is much too late.  I imagine it may take a day or two to get in touch with the sales representative, so I really think we should give at least a one day notice.  Unless we get direction otherwise from @nedrichards, I would add one more notification for ``60 * 60 * 24, // 1 day` and then add string processing for number of days.

File: js/misc/paygManager.js [hunk 2]
```diff
@@ -88,6 +110,27 @@ function _isPaygEnabled() {
     return false;
 }
 
+// Takes an UNIX timestamp (in seconds) and returns a string
+// with a precision level appropriate to show to the user.
+function timeToString(seconds) {
+    if (seconds < 60)
+        return Gettext.ngettext("%s second", "%s seconds", seconds).format(seconds);
+
+    let minutes = Math.ceil(seconds / 60);
```
[the remark this answers is not in the document]
Comment (rshuler) [reply js/misc/paygManager.js]: Actually, I see that we explicitly call with the round number of target seconds, so we should do the proper rounding and use floor here, especially since when we call this upon logging in, it won't be with the target number of seconds but with the actual time remaining, so we do need to display an amount of time that does not exceed what is actually left.

File: js/misc/paygManager.js [hunk 3]
```diff
@@ -88,6 +110,27 @@ function _isPaygEnabled() {
...
+
+    let hours = Math.floor(minutes / 60);
+    let minutesPast = minutes % 60;
+
+    if (minutesPast == 0)
+        return Gettext.ngettext("%s hour", "%s hours", hours).format(hours);
+
+    // For anything that is not a round number of hours, we resort to the
+    // number of minutes, which is simpler than mixing hours and minutes.
+    return "%s minutes".format(minutes);
```
Comment (rshuler) [inline js/misc/paygManager.js]: Given my suggestion to add a one-day warning, the user could boot with something like 15 hours and 39 minutes.  In that case, we should round down to 15 hours rather than display 939 minutes.  I'd be okay with showing full minutes for times less than 2 hours, but for times above 2 hours I would just round down to the nearest number of hours.

File: js/misc/paygManager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mariospr) [reply js/misc/paygManager.js]: You're right. I used ceiling because I initially had a different idea of what I ended up implementing, but I agree it should be `Math.floor()` now.

File: js/misc/paygManager.js [hunk 4]
```diff
@@ -60,6 +63,25 @@ const DBusErrorsMapping = {
...
+const NOTIFICATION_TITLE_FORMAT_STRING = _("Pay-as-you-go time will expire in %s");
+const NOTIFICATION_DETAILED_TEXT = _("Talk to your sales representative to buy a new code.")
+
+// This list defines the different instants in time where we would
+// want to show notifications to the user reminding that the payg
+// subscription will be expiring soon.
+//
+// It contains a list of integers representing the number of seconds
+// earlier to the expiration time when we want to show a notification,
+// which needs to be sorted in DESCENDING order.
```
[the remark this answers is not in the document]
Comment (mariospr) [reply js/misc/paygManager.js]: Yes, but at the same time it seems kind of useless since this is a static list we'll maintain and we can  make sure that it's sorted.

File: js/misc/paygManager.js [hunk 5]
```diff
@@ -190,6 +239,73 @@ var PaygManager = new Lang.Class({
...
+        if (targetAlertTime == 0)
+            return;
+
+        this._expirationReminderId = Mainloop.timeout_add_seconds(secondsLeft - targetAlertTime, () => {
+            // We want to show "round" numbers in the notification, matching
+            // whatever is specified in the notificationAlertTimeSecs array.
+            this._notifyPaygReminder(targetAlertTime);
+            this._updateExpirationReminders();
+
+            this._expirationReminderId = 0;
```
[the remark this answers is not in the document]
Comment (mariospr) [reply js/misc/paygManager.js]: Good catch. I didn't realize because in my mind I was setting the ID to `0` at the time I was removing the source (by returning `G_SOURCE_REMOVE`), but the fact that this is calling `_updateExpirationReminders()` again causes the problem you mention.

I'll change that.

File: js/misc/paygManager.js [hunk 6]
```diff
@@ -272,6 +273,16 @@ var PaygManager = new Lang.Class({
         });
     },
 
+    _maybeNotifyUser: function() {
+        // Sanity check.
+        if (notificationAlertTimesSecs.length == 0)
+            return;
+
+        let secondsLeft = this._timeRemainingSecs();
+        if (this._timeRemainingSecs() <= notificationAlertTimesSecs[0])
```
[the remark this answers is not in the document]
Comment (mariospr) [reply js/misc/paygManager.js]: Thanks, I'll fix those too.

File: js/misc/paygManager.js (hunk 1 above)
Comment (mariospr) [reply js/misc/paygManager.js]: I'll add for **1 and 2 days** (based on your comment on phabricator, which I agree with), and change the string processing function.

File: js/misc/paygManager.js (hunk 3 above)
Comment (mariospr) [reply js/misc/paygManager.js]: Sounds good, I'll change that then.

File: js/misc/paygManager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mariospr) [reply js/misc/paygManager.js]: Actually, we can't plainly use `floor()` either in all cases, as I've realized while testing the latest changes I just made, which show an issue when you **unlock the machine from outside a session**:

Imagine the scenario when you boot up to a locked machine and unlock it with a code for 2 days, and then you log in to the recently-unlocked system: at the time the shell starts the timestamp that will be checked to decide whether to show a notification to the user with the time remaining will already be different (more recent) than the one used when unlocking the machine, meaning that the string formatting logic (that uses `floor()` to calculate the number of minutes, hours and days) will makes up end up showing a message like "PAYG time will expire in 1 day" to the user, which would be very confusing since you just entered a code for 2 days!

In the same line, if you enter a code for 1 day, you'll start to a notification saying "PAYG time will expire in 23 hours", which is not as bad but still a bit confusing, although in that case I think it cold be much less confusing than in the case of 2 vs 1 days, although I still think some correction might be needed there as well.

Last, if you enter a code for 1 hour, you'll boot up to a message that says "PAYG time will expire in 59 minutes", which I don't see any trouble with in terms of causing confusion.

TL;DR: I think we should add some kind of epsilon before using `floor()` directly for some of the biggest units, to avoid confusion. In particular, I think we should definitely use an EPSILON for the case of days because it's **very** confusing to unlock with a code for 2 days and see right away that you only have one left. For strings formatted in hours ("23 hours left" if you just unlocked for 1 day), I'm not really that concerned but I think some adjustment would be useful too.

[the remark this answers is not in the document]
Comment (mariospr) [reply js/misc/paygManager.js]: On a second thought, I'm not entirely happy with this solution either, since it doesn't really removes potential confusion from the user in cases other than "logging in just (in less than 5 min) after unlocking the machine. For instance, imagine you unlock the machine with a code for 2 days but take more than 5 min to log-in for any reason: in that case we'd be still showing "1 day left" which might be equally confusing. Or maybe you log in fast enough but then reboot a few minutes later and you see "1 day left".... confusing.

So, I now think a better solution might actually be something not involving epsilons, but simply reporting the actual amount of time with a bit more of detail, that is: returning a string "in seconds" for times under 1 minute, a string "in minutes" for times under 2 hours, "in hours and minutes" (if applicable) for times under 1 day, and then "in days and hours" (if applicable) for anything longer than that in days.

I've tested this and I think that's the approach that works best, and seeing  "1 day 23 hours left" right after unlocking with a code for 2 days should not be much of a problem, while way less confusing than just "flooring" it to "1 day". Also, note that the periodic reminders will be always round numbers (e.g. "2 days", "1 hour"...) because we define them that way, and so the more detailed format will only be shown as a one-off notification on startup (where you would expect a bit more of precision),  so I think it should be fine.

[the remark this answers is not in the document]
Comment (rshuler) [reply js/misc/paygManager.js]: Another possibility is to avoid the crazy rounding altogether and display something like the following...

Your code will expire in:
XX day(s) YY hour(s) ZZ minute(s)

File: js/misc/paygManager.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (rshuler) [reply js/misc/paygManager.js]: Yeah, that's fine.

File: js/misc/paygManager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mariospr) [reply js/misc/paygManager.js]: Yes, I thought of that too. But the problem is that we define the periodic notifications with such values that you would print many "zero values" in there (e.g. "0 days, 2 hours 0 minutes"), so some formatting is necessary.

File: js/misc/paygManager.js [hunk 7]
```diff
@@ -60,6 +63,27 @@ const DBusErrorsMapping = {
     DISABLED          : 'com.endlessm.Payg1.Error.Disabled',
 };
 
+// Title and description text to be shown in the periodic reminders.
+const NOTIFICATION_TITLE_FORMAT_STRING = _("Pay-as-you-go time will expire in %s");
```
Comment (rshuler) [inline js/misc/paygManager.js]: I did not notice this before, and I'm not sure if "Pay-as-you-go" is intended to be a user-facing name of the feature.  That said, I'll merge as is for now, but there is a chance that we may have to tweak that string.

File: js/misc/paygManager.js [hunk 8]
```diff
@@ -88,6 +112,57 @@ function _isPaygEnabled() {
     return false;
 }
 
+// Takes an UNIX timestamp (in seconds) and returns a string
+// with a precision level appropriate to show to the user.
+//
+// The returned string will be formatted just in seconds for times
```
Comment (rshuler) [inline js/misc/paygManager.js]: I'm okay with this decision for now, though I still would consider simplifying to just always show "XX days YY hours ZZ minutes".

File: js/misc/paygManager.js [hunk 9]
```diff
@@ -88,6 +112,57 @@ function _isPaygEnabled() {
...
+// under 1 minute, in minutes for times under 2 hours, in hours and
+// minutes (if applicable) for times under 1 day, and then in days
+// and hours (if applicable) for anything longer than that in days.
+//
+// Some examples:
+//   - 45 seconds => "45 seconds"
+//   - 60 seconds => "1 minute"
+//   - 95 seconds => "1 minute"
+//   - 120 seconds => "2 minutes"
```
Comment (rshuler) [inline js/misc/paygManager.js]: For this case, I would probably follow the same rule that you use for showing number of minutes under 2 hours, and show the full number of seconds if under 2 minutes.  But I don't feel that strongly, so I'm going to merge as is.

File: js/misc/paygManager.js (hunk 7 above)
Comment (mariospr) [reply js/misc/paygManager.js]: I took that from Robin, and it's also consistent with the designs in invision (see T21811), but I also share your concern.
